### PR TITLE
[uninstall] ensure service is stopped on windows

### LINKED ISSUE #4164: [Flaky Test]: TestLogIngestionFleetManaged – Failed to remove install directory
### Failing test case

TestLogIngestionFleetManaged

### Error message

Failed to remove install directory

### Build

https://buildkite.com/elastic/elastic-agent/builds/6667#018d5f15-e83c-4be2-80e2-64c2d17dcfa0

### OS

Windows

### Stacktrace and notes

```markdown
fixture_install.go:215: 
        	Error Trace:	C:/Users/windows/agent/pkg/testing/fixture_install.go:215
        	            				C:/Program Files/Go/src/testing/testing.go:1169
        	            				C:/Program Files/Go/src/testing/testing.go:1347
        	            				C:/Program Files/Go/src/testing/testing.go:1589
        	Error:      	Received unexpected error:
        	            	error running uninstall command: exit status 1
        	Test:       	TestLogIngestionFleetManaged
        	Messages:   	uninstalling agent failed. Output: "
[    ] Stopping service  [0s] 
                              

[  ==] Successfully stopped service  [0s] 
                                          

[ ===] Stopping upgrade watcher; none found  [0s] 
                                                  

[=   ] Removing service  [0s] 
                                                  

[=   ] Successfully uninstalled service  [0s] 
                                                  

[=   ] Removing install directory  [0s] 
                                                  

[====] Failed to remove install directory  [10s] 
                                                  

[====] Failed to uninstall agent  [10s] Error uninstalling.  Printing logs
{debug 2024-01-31 11:00:03.7594074 +0000 GMT m=+0.396454701 processes Error fetching PID info for 0, skipping: GetInfoForPid: could not get all information for PID 0: error fetching name: OpenProcess failed for pid=0: The parameter is incorrect.
error fetching status: OpenProcess failed for pid=0: The parameter is incorrect. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:173 }
{debug 2024-01-31 11:00:03.7603068 +0000 GMT m=+0.397354101 processes Error fetching PID info for 4, skipping: GetInfoForPid: could not get all information for PID 4: error fetching name: GetProcessImageFileName failed for pid=4: GetProcessImageFileName failed: invalid argument /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:173 }
{debug 2024-01-31 11:00:03.7603938 +0000 GMT m=+0.397441101 processes Non fatal error fetching PID some info for 116, metrics are valid, but partial: FillMetricsRequiringMoreAccess: error fetching process args: Not enough privileges to fetch information: OpenProcess failed: Access is denied. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:176 }
{debug 2024-01-31 11:00:03.7613448 +0000 GMT m=+0.398392101 processes Non fatal error fetching PID some info for 364, metrics are valid, but partial: FillMetricsRequiringMoreAccess: error fetching process args: Not enough privileges to fetch information: OpenProcess failed: Access is denied. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:176 }
{debug 2024-01-31 11:00:03.7613448 +0000 GMT m=+0.398392101 processes Non fatal error fetching PID some info for 476, metrics are valid, but partial: FillMetricsRequiringMoreAccess: error fetching process args: Not enough privileges to fetch information: OpenProcess failed: Access is denied. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:176 }
{debug 2024-01-31 11:00:03.7623309 +0000 GMT m=+0.399378201 processes Non fatal error fetching PID some info for 560, metrics are valid, but partial: FillMetricsRequiringMoreAccess: error fetching process args: Not enough privileges to fetch information: OpenProcess failed: Access is denied. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:176 }
{debug 2024-01-31 11:00:03.7623309 +0000 GMT m=+0.399378201 processes Non fatal error fetching PID some info for 568, metrics are valid, but partial: FillMetricsRequiringMoreAccess: error fetching process args: Not enough privileges to fetch information: OpenProcess failed: Access is denied. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:176 }
{debug 2024-01-31 11:00:03.7632954 +0000 GMT m=+0.400342701 processes Non fatal error fetching PID some info for 700, metrics are valid, but partial: FillMetricsRequiringMoreAccess: error fetching process args: Not enough privileges to fetch information: OpenProcess failed: Access is denied. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:176 }
{debug 2024-01-31 11:00:03.7921051 +0000 GMT m=+0.429152401 processes Non fatal error fetching PID some info for 3104, metrics are valid, but partial: FillMetricsRequiringMoreAccess: error fetching process args: Not enough privileges to fetch information: OpenProcess failed: Access is denied. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:176 }
{debug 2024-01-31 11:00:03.7968609 +0000 GMT m=+0.433908201 processes Non fatal error fetching PID some info for 4920, metrics are valid, but partial: FillMetricsRequiringMoreAccess: error fetching process args: Not enough privileges to fetch information: OpenProcess failed: Access is denied. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:176 }
{debug 2024-01-31 11:00:03.8087051 +0000 GMT m=+0.445752401 processes Non fatal error fetching PID some info for 2508, metrics are valid, but partial: FillMetricsRequiringMoreAccess: error fetching process args: Not enough privileges to fetch information: OpenProcess failed: Access is denied. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:176 }
{debug 2024-01-31 11:00:03.8096899 +0000 GMT m=+0.446737201 processes Non fatal error fetching PID some info for 4748, metrics are valid, but partial: FillMetricsRequiringMoreAccess: error fetching process args: Not enough privileges to fetch information: OpenProcess failed: Access is denied. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:176 }
{debug 2024-01-31 11:00:03.8223729 +0000 GMT m=+0.459420201 processes Non fatal error fetching PID some info for 388, metrics are valid, but partial: FillMetricsRequiringMoreAccess: error fetching process args: Not enough privileges to fetch information: OpenProcess failed: Access is denied. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:176 }
{debug 2024-01-31 11:00:03.8234031 +0000 GMT m=+0.460450401 processes Non fatal error fetching PID some info for 5136, metrics are valid, but partial: FillMetricsRequiringMoreAccess: error fetching process args: Not enough privileges to fetch information: OpenProcess failed: Access is denied. /go/pkg/mod/github.com/elastic/elastic-agent-system-metrics@v0.9.1/metric/system/process/process.go:176 }
{warn 2024-01-31 11:00:03.9561501 +0000 GMT m=+0.593197401 cfgwarn DEPRECATED: Treating the CommonName field on X.509 certificates as a host name when no Subject Alternative Names are present is going to be removed. Please update your certificates if needed. Will be removed in version: 8.0.0 /go/pkg/mod/github.com/elastic/elastic-agent-libs@v0.7.4/transport/tlscommon/config.go:101 }
{debug 2024-01-31 11:00:03.957143 +0000 GMT m=+0.594190301 state_migration state store C:\\Program Files\\Elastic\\Agent\\data\\elastic-agent-86745f\\state.enc already exists /go/src/github.com/elastic/elastic-agent/internal/pkg/agent/storage/store/state_store.go:159 }
{debug 2024-01-31 11:00:04.0040017 +0000 GMT m=+0.641049001 composable Starting controller for composable inputs /go/src/github.com/elastic/elastic-agent/internal/pkg/composable/controller.go:112 }
{debug 2024-01-31 11:00:04.0040017 +0000 GMT m=+0.641049001 composable Started controller for composable inputs /go/src/github.com/elastic/elastic-agent/internal/pkg/composable/controller.go:155 }
{debug 2024-01-31 11:00:04.0040017 +0000 GMT m=+0.641049001 docker Docker client will negotiate the API version on the first request. /go/pkg/mod/github.com/elastic/elastic-agent-autodiscover@v0.6.6/docker/client.go:49 }
{info 2024-01-31 11:00:04.0040017 +0000 GMT m=+0.641049001 composable.providers.docker Docker provider skipped, unable to connect: protocol not available /go/src/github.com/elastic/elastic-agent/internal/pkg/composable/providers/docker/docker.go:44 }
{debug 2024-01-31 11:00:04.0040017 +0000 GMT m=+0.641049001 composable Variable state changed for composable inputs; debounce started /go/src/github.com/elastic/elastic-agent/internal/pkg/composable/controller.go:191 }
{debug 2024-01-31 11:00:04.0048934 +0000 GMT m=+0.641940701 composable kubernetes_secrets provider skipped, unable to connect: unable to build kube config due to error: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable /go/src/github.com/elastic/elastic-agent/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:81 }
{debug 2024-01-31 11:00:04.0048934 +0000 GMT m=+0.641940701 composable.providers.kubernetes Kubernetes provider for resource pod skipped, unable to connect: unable to build kube config due to error: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable /go/src/github.com/elastic/elastic-agent/internal/pkg/composable/providers/kubernetes/kubernetes.go:106 }
{debug 2024-01-31 11:00:04.0048934 +0000 GMT m=+0.641940701 composable.providers.kubernetes Kubernetes provider for resource node skipped, unable to connect: unable to build kube config due to error: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable /go/src/github.com/elastic/elastic-agent/internal/pkg/composable/providers/kubernetes/kubernetes.go:106 }
{debug 2024-01-31 11:00:04.0048934 +0000 GMT m=+0.641940701 composable Kubernetes leaderelection provider skipped, unable to connect: unable to build kube config due to error: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable /go/src/github.com/elastic/elastic-agent/internal/pkg/composable/providers/kubernetesleaderelection/kubernetes_leaderelection.go:53 }
{debug 2024-01-31 11:00:04.1095056 +0000 GMT m=+0.746552901 composable Computing new variable state for composable inputs /go/src/github.com/elastic/elastic-agent/internal/pkg/composable/controller.go:207 }
{debug 2024-01-31 11:00:04.1095056 +0000 GMT m=+0.746552901 composable Stopping controller for composable inputs /go/src/github.com/elastic/elastic-agent/internal/pkg/composable/controller.go:159 }
{debug 2024-01-31 11:00:04.2198811 +0000 GMT m=+0.856928401 composable Stopped controller for composable inputs /go/src/github.com/elastic/elastic-agent/internal/pkg/composable/controller.go:188 }
Error: error uninstalling agent: failed to remove installation directory (C:\\Program Files\\Elastic\\Agent): timed out while removing \"C:\\\\Program Files\\\\Elastic\\\\Agent\". Last error: remove C:\\Program Files\\Elastic\\Agent\\data\\elastic-agent-86745f\\elastic-agent.exe: Access is denied.
For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.13/fleet-troubleshooting.html
"
--- FAIL: TestLogIngestionFleetManaged (182.71s)
```

Possibly related to https://github.com/elastic/elastic-agent/issues/4143
Possible duplicate of https://github.com/elastic/elastic-agent/issues/4092

### PR DESCRIPTION
**Pull request by leehinman**
## What does this PR do?

On Windows, this change will poll the Windows Service Manager every 500 msec, for up to 30 seconds, to make sure the service is actually stopped before attempting to remove the installed directory.

## Why is it important?

The kardianos service manager doesn't distinguish between 'Stopped' and 'StopPending' on Windows, so we need to query to make sure the service is really stopped.  Otherwise we can try to remove the files while the service is still running.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test

## Author's Checklist

- [ ]

## How to test this PR locally

Install & uninstall the agent on Windows and Linux.

The integration tests do this as well.

## Related issues

- Closes #4164

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

File: internal/pkg/agent/install/wait_service_windows.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	defer ticker.Stop()
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			status, err := s.Query()
+			if err != nil {
+				return fmt.Errorf("error querying service (%s): %w", service, err)
```
Comment (pchila) [inline internal/pkg/agent/install/wait_service_windows.go:45]: I am not sure we should abandon the wait loop if we fail to query. If we an error is returned by `s.Query()` is `s` still usable ?
 Just thinking that if we can reuse the object and we have the time we could keep going and try to query again (logging the failure in the loop anyways)

**Review by pchila — COMMENTED** [review]
Just a question about abandoning the wait loop on windows if we fail to query the status once. The rest of the code seems ok.

I am not sure how easy it would be to have a unit test (those run without admin privileges though) to launch a small "sleepy" service that takes 5-10 seconds to shut down to verify that this works....

File: internal/pkg/agent/install/wait_service_windows.go (hunk 1 above)
Comment (leehinman) [reply internal/pkg/agent/install/wait_service_windows.go:45]: yeah, I was considering this.  But I looked at the errors that can be returned, https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-queryservicestatusex and those don't look like transient errors, so I was thinking we should probably just give up.

**Comment by leehinman** [comment]
> I am not sure how easy it would be to have a unit test (those run without admin privileges though) to launch a small "sleepy" service that takes 5-10 seconds to shut down to verify that this works....

You have to create the service in the service manager, and that requires elevated privileges, so we would really need to do an integration test.  I'm not too worried about this since the reason we found it was existing tests cause it to happen on a pretty regular basis.

**Review by pchila — APPROVED** [review]

**Comment by leehinman** [comment]
@cmacknz I think this will need a force merge (to get past SonarQube), unless we want to try an write an integration test for it that fails all the time, but that seems like overkill since we have integration tests that seem to hit it already.

**Comment by cmacknz** [comment]
> unless we want to try an write an integration test for it that fails all the time, but that seems like overkill since we have integration tests that seem to hit it already.

I am considering this. We have had enough problems here that it may be worth writing a small stress test that just installs and uninstalls the agent over and over as fast as it can.

The integration tests do cover this code, but at the same time if we wrote the stress test the first time maybe we'd have caught this problem there.

Thoughts?

File: internal/pkg/agent/install/wait_service_windows.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+			}
+			if status.State == svc.Stopped {
+				return nil
+			}
+		case <-timer.C:
+			return fmt.Errorf("timed out waiting for service (%s) to stop", service)
```
[the remark this answers is not in the document]
Comment (AndersonQ) [reply internal/pkg/agent/install/wait_service_windows.go:51]: [Suggestion]
Also you could save the last status and also add it to the error. It'd help us to understand what happened if it times out.

**Comment by leehinman** [comment]
> > unless we want to try an write an integration test for it that fails all the time, but that seems like overkill since we have integration tests that seem to hit it already.
> 
> I am considering this. We have had enough problems here that it may be worth writing a small stress test that just installs and uninstalls the agent over and over as fast as it can.
> 
> The integration tests do cover this code, but at the same time if we wrote the stress test the first time maybe we'd have caught this problem there.
> 
> Thoughts?

Install/Uninstall for some fixed number of iterations seems reasonable to me.  My only concern is with time, the integrations tests already take a fair amount of time.

How about I write a test now that does something like 100 install/uninstall iterations?  And then when the long running tests are merged we can move it there and run 10k or 100k iterations.

**Comment by cmacknz** [comment]
> How about I write a test now that does something like 100 install/uninstall iterations? And then when the long running tests are merged we can move it there and run 10k or 100k iterations.

Sounds reasonable to me.

**Comment by cmacknz** [comment]
This should have a changelog entry.

File: internal/pkg/agent/install/wait_service_windows.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	"fmt"
+	"time"
+
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/mgr"
+)
+
+// isStopped queries the Windows service manager to see if the state of the service is stopped.
+// it will repeat the query every every 'interval' until the 'timeout' is reached.  It returns
+// nil if the system is stopped within the timeout period, and an error if it isn't.
```
[the remark this answers is not in the document]
Comment (leehinman) [reply internal/pkg/agent/install/wait_service_windows.go]: fixed type and added to description.

File: internal/pkg/agent/install/wait_service_windows.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (leehinman) [reply internal/pkg/agent/install/wait_service_windows.go]: added both suggestions

File: testing/integration/install_test.go [hunk 4]
```diff
@@ -167,6 +168,68 @@ func TestInstallWithBasePath(t *testing.T) {
...
+		// installs Elastic Agent.
+		Local: false,
+	})
+
+	maxRunTime := 2 * time.Minute
+	iterations := 100
+	for i := 0; i < iterations; i++ {
+		t.Run(fmt.Sprintf("%s-%d", t.Name(), i), func(t *testing.T) {
+
+			var defaultBasePath string
```
Comment (cmacknz) [inline testing/integration/install_test.go:193]: nit: We have this logic in 3 other places so it could probably become a helper function

**Review by cmacknz — APPROVED** [review]
LGTM except the missing changelog.

**Comment by leehinman** [comment]
> Currently the TestFleetManagedUpgradePrivileged test on this PR is failing with:

I'm pretty sure this is because `elastic-agent-8.13.0-SNAPSHOT-c09911` doesn't have this change.

**Comment by cmacknz** [comment]
Force merging to bypass the test coverage requirement, the changes are covered by a new uninstall stress integration test.

**Comment by cmacknz** [comment]
The 8.12 branch is seeing failures with the following that might be related to this change.

```
[    ] Failed to uninstall current Elastic Agent  [0s] Error uninstalling.  Printing logs
        	            	Error: error installing package: failed to uninstall Agent at (C:\Program Files\Elastic): failed to issue stop service (Elastic Agent): The service cannot accept control messages at this time.
        	            	For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.12/fleet-troubleshooting.html
        	            	]
```

https://buildkite.com/elastic/elastic-agent/builds/7231#018dade7-b360-4571-b713-e7963dca3de7 and https://buildkite.com/elastic/elastic-agent/builds/7226

**Pull request merged**